rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,135 @@
+
+## What
+
+Users can change their permissions in a way that leaves them incapable of making further changes to the cluster - in effect they lock themselves out.
+
+Currently, there are multiple ways to trigger this scenario:
+
+| initial state                              | change                                                                                      |
+| ------------------------------------------ | ------------------------------------------------------------------------------------------- |
+| 1 _admin_ and _regular users_              | _admin_ removes himself from an `editor` role                                               |
+| 1 _admin_ and _regular users_              | _admin_ loses access to own account                                                         |
+| 1 _admin_, _SSO admin_ and _regular users_ | _SSO admin_ removes the `editor` role from _admin_ and SSO connector. _SSO admin_ logs out. |
+| _SSO admin_ and _regular users_            | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_      |
+
+Teleport already protects against a similar case in the following scenario:
+
+| initial state               | change                                                                             |
+| --------------------------- | ---------------------------------------------------------------------------------- |
+| _admin_ and _regular users_ | _admin_ tries to delete his account. Operation is not permitted, an error is shown |
+
+## Terminology
+
+- _admin_: local user with `editor` role
+- _SSO admin_: user with `editor` role added via SSO connector (eg. GitHub)
+- _regular user_: user without `editor` role
+
+## Why
+
+When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+Further, for our Cloud offering we don't perform account resets as a policy. The only recurse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data-loss.
+
+On-prem customers could work around this "locked out" state by manually instrumenting the auth process using tctl.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can be `editors`.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least 2 local _admins_
+
+This will ensure that:
```

Comment:
> preventing the deletion the last role with create/update rules to the role resource.

We already prevent deletion of any role which is currently assigned to a user, I don't think we should need any extra check for this.

> preventing the removal of create/update rules from the last role that's got them.

To be more precise, if there is role `admin` and `admin-nobody-has-this-role`, and both have create/update role rules, we still need to prevent the removal of create/update rules from `admin`. The other role with the requisite rules still exists, but there is no user who is assigned it